### PR TITLE
Fix: adapt logs handler to jobs

### PR DESCRIPTION
**Pull request by SoniaGrh**
* Handle logs using only job's methods (and not experiment's)

* Handle progress bar in real time (logging progression at the same line)

* Create a LogMonitor class to monitor the log of one job

* Create a LogTailer class to handle each line of a log file

* Add a utils file to get the Picsellia client and job

### CODE REVIEW

**Comment by picsalex** [comment]
I would personally slightly change the folder hierarchy for the `base/` folder, something like :

```
base/
|-- logs/
|   |-- __init__.py
|   |-- handler.py (previously logs_handler.py)
|   |-- monitor.py (previously log_monitor.py)
|   |-- tailer.py (previously log_tailer.py)
|
|-- utils/
|   |-- __init__.py
|   |-- picsellia.py (i guess this should conflict with the picsellia package by calling utils.picsellia)
```

File: base/common/common_utils/log_monitor.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+from .log_tailer import LogTailer
+from .picsellia_utils import (
+    get_picsellia_job,
+    get_picsellia_client,
+    get_picsellia_experiment,
+)
+
+
+def starts_buffer(line: str) -> Match[str] | None:
```
Comment (picsalex) [inline base/common/common_utils/log_monitor.py:16]: I would put the variables inside constants, and rename the functions to better match their purpose, something like:

```python
# Constants for better clarity and avoiding magic strings
BUFFER_START_PATTERN = re.compile(r"--[0-9]--")
BUFFER_END_PATTERN = re.compile(r"---[0-9]---")

def does_buffer_start(line: str) -> Match[str] | None:
    """Check if line starts a buffer."""
    return BUFFER_START_PATTERN.match(line[:6])


def does_buffer_end(line: str) -> Match[str] | None:
    """Check if line ends a buffer."""
    return BUFFER_END_PATTERN.match(line[:8])
```

File: base/common/common_utils/log_monitor.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+        self.logs = {}
+
+        self.buffer = []
+        self.start_buffer = False
+        self.buffer_length = 0
+
+        self.progress_line_nb = 0
+
+    def handle_exit_code(self, line: str) -> None:
```
Comment (picsalex) [inline base/common/common_utils/log_monitor.py:42]: I would also put this into a constant, like:

```python
EXIT_CODE_PATTERN = re.compile(r"--ec-- ([0-9]+)")

 def handle_exit_code(self, line: str) -> None:
        exit_match = EXIT_CODE_PATTERN.search(line)
        if exit_match:
            self.end_log_monitoring(int(exit_match.group(1)), "--#--End job")

```

File: base/common/common_utils/log_tailer.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        line (str): The line to be formatted. It's assumed that this is not None.
+
+    Returns:
+        str: The formatted line. If the input line ended with one or more backspaces, these are removed.
+        If the input line did not end with a newline, one is added.
+    """
+    return line.rstrip("\n").rstrip(chr(8)) + "\n"
+
+
+def is_progress_line(line: str) -> bool:
```
Comment (picsalex) [inline base/common/common_utils/log_tailer.py:38]: can be refactored into something easier to read/understand/maintain:

```python
symbols = ["%", "<", "[", "]"]
return all(symbol in line for symbol in symbols)
```

**Comment by SoniaGrh** [comment]
> I would personally slightly change the folder hierarchy for the `base/` folder, something like :
> 
> ```
> base/
> |-- logs/
> |   |-- __init__.py
> |   |-- handler.py (previously logs_handler.py)
> |   |-- monitor.py (previously log_monitor.py)
> |   |-- tailer.py (previously log_tailer.py)
> |
> |-- utils/
> |   |-- __init__.py
> |   |-- picsellia.py (i guess this should conflict with the picsellia package by calling utils.picsellia)
> ```

The folder "utils" was also conflicting with the utils folder/file of training images, should I replace it with
> |-- base_utils/
> |   |-- __init__.py
> |   |-- picsellia_utils.py 

Or do you have a better suggestion  ?

**Review by picsalex — APPROVED** [review]

**Pull request merged**